src/emu/http.cpp

```diff
@@ -427,9 +420,29 @@ bool http_manager::read_file(std::ostream &os, const std::string &path) {
 void http_manager::serve_document(http_request_ptr request, http_response_ptr response, const std::string &filename) {
 	if (!m_active) return;
 
+#if DEBUG
+	std::cerr << "serve_document('" << filename << "')" << std::endl;
+#endif
+
+	if (!is_path_safe(filename)) {
+#if DEBUG
+		std::cerr << "serve_document(): filename is not safe, serving NotFound" << std::endl;
+#endif
+		this->serve_document(request, response, "NotFound.html");
+		return;
+	}
+
 	std::ostringstream os;
 	if (read_file(os, filename))
 	{
+		std::size_t last_slash_pos = filename.find_last_of("/");
+		std::size_t last_dot_pos = filename.find_last_of(".");
```

Comment:
You can restrict this search to the position of the last slash if found.

Reply:
Done.